bee-vote/src/context.rs

```diff
@@ -0,0 +1,252 @@
+    }
+}
+
+impl VoteObject {
+    /// Get a `MessageId` if this is a Timestamp object.
+    /// Returns `Some(id)` if this is the case, `None` if not.
+    pub fn message_id(&self) -> Option<MessageId> {
+        if let Self::Timestamp(id) = *self {
+            Some(id)
+        } else {
+            None
+        }
+    }
+
+    /// Get a `TransactionId` if this is a Conflict object.
+    /// Returns `Some(id)` if this is the canse, `None` if not.
+    pub fn transaction_id(&self) -> Option<TransactionId> {
+        if let Self::Conflict(id) = *self {
+            Some(id)
+        } else {
+            None
+        }
+    }
+}
+
+/// Builder pattern struct for instantiating a `VoteContext`.
+pub struct VoteContextBuilder {
+    /// Object of the voting (conflict or timestamp), and associated ID.
+    object: VoteObject,
+    /// Opinions held by this context on the vote object.
+    opinions: Option<Opinions>,
+}
+
+impl VoteContextBuilder {
+    /// Create a new `VoteContextBuilder`, defining an ID and an `ObjectType` (voting object).
+    pub fn new(object: VoteObject) -> Self {
+        Self { object, opinions: None }
+    }
+
+    /// Set a single initial `Opinion`.
+    pub fn with_initial_opinion(mut self, opinion: Opinion) -> Self {
+        self.opinions = Some(Opinions::new(vec![opinion]));
+        self
+    }
+
+    /// Set a list of initial `Opinion`s.
+    pub fn with_initial_opinions(mut self, opinions: Opinions) -> Self {
+        self.opinions = Some(opinions);
+        self
+    }
+
+    /// Build a `VoteContext`.
+    /// Note: this will panic if no initial opinions have been provided.
+    pub fn build(self) -> Result<VoteContext, Error> {
+        Ok(VoteContext {
+            object: self.object,
+            liked: LIKED_INITIAL,
+            rounds: 0,
+            opinions: self.opinions.ok_or(Error::NoInitialOpinions)?,
+        })
+    }
+}
+
+/// Voting context.
+#[derive(Debug, Clone)]
+pub struct VoteContext {
+    /// Object type of the vote and related object ID.
+    object: VoteObject,
+    /// The percentage of `OpinionGiver`s who liked this item on the last query.
+    liked: f64,
+    /// The number of voting rounds performed so far.
+    rounds: u32,
+    /// List of opinions formed at the end of each voting round.
+    /// The first in this list is the initial opinion when this `VoteContext` was created.
+    opinions: Opinions,
+}
+
+impl VoteContext {
+    /// Constructs a new `VoteContext`.
+    pub(crate) fn new(object: VoteObject, initial_opinion: Opinion) -> Self {
+        Self {
+            object,
+            liked: LIKED_INITIAL,
+            rounds: 0,
+            opinions: Opinions::new(vec![initial_opinion]),
+        }
+    }
+
+    /// Add the given `Opinion` to the `VoteContext`.
+    pub(crate) fn add_opinion(&mut self, opinion: Opinion) {
+        self.opinions.push(opinion);
+    }
+
+    /// Retrieve the last formed `Opinion`.
+    pub fn last_opinion(&self) -> Option<Opinion> {
```

Comment:
Out of curiosity, is there any situation where it's valid not to have any opinions at all? The `new()` function above enforces that an initial opinion is provided (guaranteed initial opinion), and the `VoteContextBuilder` enforces that the caller uses either `with_initial_opinion` (guaranteed initial opinion)  or `with_initial_opinions` (possible to construct an `Opinions` with no elements).

If it is indeed the case that there **must** be an initial opinion, the VoteContextBuilder could be updated to check the length of the opinions vector and error otherwise (also with `NoInitialOpinions`). This way the API here could be simplified to just return an `Opinion`.

The downside of this is that if there somehow became a way to construct a `VoteContext` that bypassed these checks, the only option here would be to either: a) return a default opinion (is there a "safe" one, protocol wise?) or b) panic (also not great). It seems like the risk here is pretty low though since there are only two possible ways to construct a `VoteContext`.

WDYT?

Reply:
It is valid for a voter to form no `Opinion`s on an object. These will get skipped in the "liked" calculation at [this line](https://github.com/iotaledger/bee/blob/f1726f92e705a8393be6715b79387aeeef4374ce/bee-vote/src/fpc.rs#L467). What I have realised because of this comment, though, is that the `min_opinions_received` field should be restricted to `> 0`, to stop any nasty panics if a user mistakenly sets it to 0 (this would skip the continue statement and cause a division by zero at [this line](https://github.com/iotaledger/bee/blob/f1726f92e705a8393be6715b79387aeeef4374ce/bee-vote/src/fpc.rs#L474)).

I'll address that!